### DISCUSSION TITLE
Master Password and 2FA

### DISCUSSION THREAD

**Discussion by devNull2and1** [mRemoteNG/mRemoteNG#2572] [Ideas]
first of all, thank you very much for the great program that we have been using for a long time now and are always happy when there are nightly builds. 

with increasing cyber attacks and a world, sinking into chaos, would it be possible to add a master password and optionally a two-factor authentication (challenge response with yubikey, alternative key files for usb stick)?

this way mRemoteNG would not be directly accessible if there is a threat of compromise

**Comment by devNull2and1** [comment]
@Kvarkas @xRushG would be very nice to hear what you think about that idea. at least, i would donate a yubikey for testing? :O)

**Reply by xRushG** [reply]
I have actually thought about that too. The "simple" password protection for mRemoteNG combined with an encrypted connection file is a useful feature to prevent theft of a file or the opening of that file. (But still not suitable for businesses. More for individual users.)


Multi-Factor Authentication (MFA) is a tool to protect accounts. This means MFA starts with the Windows login before launching mRemoteNG. If I want to secure my systems with MFA, it needs to be implemented for each remote system, not just for the tool used to establish connections.


In the private sphere, MFA might be considered more of a luxury since MFA software typically incurs costs and requires software and servers to secure both server and client systems. In the corporate setting, the added value of MFA in mRemoteNG is also absent, as storing passwords in mRemoteNG during connections is highly questionable. If there are no stored passwords (and/or usernames) in connections, what needs to be secured? IP addresses or hostnames are often not secret. If your system is compromised, you'll face more significant issues than just stolen hostnames or IP addresses.


The server security landscape has undergone significant changes in recent years. Topics like MFA, Privileged Access Workstations, AD Layered Model, and many more are now in focus for securely operating servers. In this context, specific tools or programs are not necessarily the driving force behind these advancements.


@devNull2and1

This is not meant as a confrontation, but rather a starting point for discussion. I'm open to being convinced otherwise. However, these were my initial thoughts. Let's explore further and see where the discussion leads.

I already own a YubiKey. However, for OTP (TOTP), the key isn't necessary. You can use Google Authenticator or tools like WinAuth for testing purposes.

**Reply by devNull2and1** [reply]
@xRushG in short, MFA would be overkill for private connection files or small env's. but also possible with yubikey/nitrokey smartcard piv feature. Anyway. 

- A master password and encrypted connection file would be a awesome standard feature if you look at keepass / keepassxc.
- A second factor like simple keyfiles on a usb stick would mean, just put the usb stick + password to access mRemoteNG.
- An alternativ to just put keyfiles to an usb stick is a challange-response with `HMAC SHA1 as defined in [RFC2104](https://datatracker.ietf.org/doc/html/rfc2104) (hashing)`

https://docs.yubico.com/yesdk/users-manual/application-otp/challenge-response.html

Just try keepassxc + yubikey with that. It is a nice feature.

**Reply by devNull2and1** [reply]
> In the corporate setting, the added value of MFA in mRemoteNG is also absent, as storing passwords in mRemoteNG during connections is highly questionable. If there are no stored passwords (and/or usernames) in connections, what needs to be secured?

thats true but there are usecases for systems that are not directly part of your own domain. alternative is to use things like keepass,keepassxc or vaultwarden to get your username/passwords and copy paste them to mremoteng. It's all about comfort, I think the features mentioned would be very great. comfort and security never goes hand in hand :(

**Reply by xRushG** [reply]
I agree, security and convenience rarely go hand in hand... 

If I understand @Kvarkas correctly, there are plans for updating the connection files (config files) and implementing some other changes for these user settings. Also, password protection is on the agenda. But I can't provide any additional details either. I'm just here to assist as a contributor. :)

Have a look on issue #2325

**Reply by devNull2and1** [reply]
Oh, i watched discussions but not issues. Im sorry, maybe link it. it is pretty much the same request.

**Reply by My1** [reply]
side note but big point tho, for locally stored stuff is that 2FA needs to output a static key for decryption, e.g. by having a static key file or abusing challenge response by always feeding the same challenge.

Actual more "dynamic" 2FA like TOTP for example would only work if the thing we want is remove where the origin can control the actual access to the stuff we want to access,

although if we would be doing e.g. per-credential storage especially with forced interaction (aka you have to press a yubikey) that would be a bit more resilient, also as you could only get one password at a time.

**Reply by devNull2and1** [reply]
i think conditions are the same, that's why keepass only have HMAC SHA1 as challange-response...users can decide if key press is needed or not. TOTP is not the correct choice for this usecase.

**Reply by My1** [reply]
but even then the point is that it is practically just one long extra password, just that you dont really have to type it. I am not against it, just want that the implications are well understood.

**Reply by xRushG** [reply]
I believe the point is that mRemoteNG is designed and should remain designed for both private users and businesses. In private settings, usability is crucial, often allowing passwords to be stored easily. While it would be advisable for users to set up a "Master Password" if they desire, the majority would likely not undertake the complexity and potential costs of implementing 2FA.

From a business perspective, security holds much greater importance. However, one might question why mRemoteNG needs to surpass or even compete with a separate tool. The primary focus of mRemoteNG lies in its function as a Connection Manager, not as a password vault. For instance, a project like KeePass possesses far more expertise regarding security matters, especially in password protection. KeePass would probably never intend to become a remote management tool.

Here, conflicting interests collide, and as previously described, implementing 2FA would indeed be highly demanding in terms of both complexity and maintenance. It's amusing how a 2FA feature sounds appealing, yet I probably wouldn't use it myself! :D But indeed, having a Master Password is fantastic. It provides a layer of protection, particularly for connections, and doesn't expose all the DNS names or IPs on a silver platter. As an administrator I strictly forbid saving passwords to Windows Credential Manager and mRemoteNG, except in programs explicitly designated for such purposes.

**Reply by devNull2and1** [reply]
private or business perspective. even if the (optional) second factor is a simple usb stick with keyfiles or the hardware-challenge with yubi.
when someone get the config file he is not able to open it with the master password, cause keyfiles / challenge is missing.
if i can decide: i would always prefer the password + keyfiles / challenge.

no one was talking about that mRmoteNG should surpass or compete with other tools like keepass. it was an example how other tools implement a simple second factor with offline challange-respsone without the complex totp integration.

**Reply by xRushG** [reply]
I just wanted to emphasize what @My1  said about the master key, which implies using an extra-long password. Combining a simple password for typing with a large key in the key file results in the overall password. However, for the less tech-savvy users (Stupidest Assumed User), all of this effort would be in **** if the key remains on the system. 😄

From a technical perspective, there are some adjustments are still needed before 2FA could be considered. But at the moment, I'm not deep enough into the project to be able to judge and it's not my responsibility. For sure, it would be nice to have, if possible. <3

**Reply by devNull2and1** [reply]
Would be awesome 👍 

> However, for the less tech-savvy users (Stupidest Assumed User), all of this effort would be in **** if the key remains on the system. 😄

Yes, just leave the key directly in the key lock to protect your house/car effectifly.

**Reply by My1** [reply]
Frankly i think yubikey could be interesting with per-credential encryption or heck yubikey aside any hmac-secret enabled fido2 key (which is most of them) could also be abused for that, as hmac secret more or less does the same (although granted you can backup a yubi better although mrem should not be used as primary credential storage anyway

**Comment by Kvarkas** [comment]
Just to FYI - technically changes I am try to implement are exactly to avoid storing any passwords within an app :) but, yes for home users its will be still possible to store them in a local db (app will handle it) at this moment we allow password protection of connection string but negative factor is if user forget it where is no recovery! So what you said is absolutely right but we need to discus more about in what format you perhaps expecting that to be implemented

**Reply by devNull2and1** [reply]
At the moment i just put mRemoteNG in a VeraCrypt Container with Password + Yubikey. And im fine.
Don't know where the problem is to implement a master password or keyfiles. Sorry don't know the project.
When i loose my password or yubikey, everything is gone. Yes. Thats fine. I like it that way.

You already have encryption. At this point it would be possible to add a master password.

![image](https://github.com/mRemoteNG/mRemoteNG/assets/144108510/83282bd0-2389-46d7-a251-f1c8c7b4365a)

**Reply by My1** [reply]
> At the moment i just put mRemoteNG in a VeraCrypt Container with Password + Yubikey. And im fine. Don't know where the problem is to implement a master password or keyfiles. Sorry don't know the project. When i loose my password or yubikey, everything is gone. Yes. Thats fine. I like it that way.
> 
> You already have encryption. At this point it would be possible to add a master password.
> 
> ![image](https://private-user-images.githubusercontent.com/144108510/321614448-83282bd0-2389-46d7-a251-f1c8c7b4365a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI4NTMyODAsIm5iZiI6MTcxMjg1Mjk4MCwicGF0aCI6Ii8xNDQxMDg1MTAvMzIxNjE0NDQ4LTgzMjgyYmQwLTIzODktNDZkNy1hMjUxLWYxYzhjN2I0MzY1YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxMVQxNjI5NDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMTgzZTY2ZTJhYzdjOTY5MmZhZGY2MTg1MzRmYmMyMDk5OGExMzgwZDZkMmIzZmU0ZTc5ZmVhYmRiOGI2OTQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vJXCN61BSlScR0DgiPBCZhZNhlyPZnQdhoL8bnIEubI)

what does it use for encryption? it needs to have some key, doesnt it.

one basically frictionless thing that might be neat would be encrypting with DPAPI (#2189) so you dont need a password for the program but the windows account takes care of the encryption.
